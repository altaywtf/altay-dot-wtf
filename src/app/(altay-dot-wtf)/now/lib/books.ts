import type { NowJSONBook } from './types'
import { REVALIDATE_NOWJSON_IN_SECONDS } from './constants'

type OkuCollectionResponse = {
  books: {
    id: string
    slug: string
    title: string
    subtitle: string
    authors: { name: string }[]
    thumbnail: string
  }[]
}

export const fetchBooks = async (): Promise<NowJSONBook[]> => {
  const OKU_COLLECTION_URL = 'https://oku.club/api/collections/user/altay/reading?format=json'

  const response = await fetch(OKU_COLLECTION_URL, {
    next: { revalidate: REVALIDATE_NOWJSON_IN_SECONDS },
  })

  const data = (await response.json()) as OkuCollectionResponse

  return data.books.map((okuBook) => ({
    url: `https://oku.club/book/${okuBook.slug}`,
    title: okuBook.title,
    subtitle: okuBook.subtitle,
    author: okuBook.authors.map((author) => author.name).join(', '),
    imageURL: okuBook.thumbnail,
  }))
}
