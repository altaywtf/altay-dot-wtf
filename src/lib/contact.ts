import fs from 'fs'
import { DATA_FOLDER_PATH } from 'lib/utils/fs'

export type ContactLink = {
  title: 'Email' | 'Twitter' | 'Github' | 'LinkedIn'
  url: string
}

export const getContactLinks = () =>
  JSON.parse(fs.readFileSync(`${DATA_FOLDER_PATH}/contact-links.json`).toString()) as ContactLink[]
