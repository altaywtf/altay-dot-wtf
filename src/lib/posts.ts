import fs from 'fs'
import matter from 'gray-matter'
import { DATA_FOLDER_PATH } from 'lib/utils/fs'
import { readMarkdownFile } from 'lib/utils/md'
import readingTime from 'reading-time'

type PostFrontMatter = {
  date: string
  oneliner: string
  title: string
}

export type Post = PostFrontMatter & {
  readingTime: string
  slug: string
  url: string
}

export const getPost = (slug: string) => {
  const file = readMarkdownFile(`/posts/${slug}.md`)
  const { content: markdown, data } = matter(file)
  const frontMatter = data as PostFrontMatter
  const readingTimeInMins = readingTime(markdown).minutes

  const post = {
    ...frontMatter,
    readingTime:
      readingTimeInMins <= 1
        ? '1 min read'
        : `${Math.floor(readingTimeInMins)} mins read`,
    slug,
    url: `/posts/${slug}`,
  } as Post

  return { markdown, post }
}

export const getPosts = () => {
  const NOTES_FOLDER_PATH = `${DATA_FOLDER_PATH}/posts`

  const slugs = fs
    .readdirSync(NOTES_FOLDER_PATH, 'utf-8')
    .filter((f) => f.endsWith('.md'))
    .map((f) => f.split('.md'))
    .map(([fileName]) => fileName)

  return slugs
    .map(getPost)
    .map((n) => n.post)
    .sort((a, b) => (Date.parse(a.date) > Date.parse(b.date) ? -1 : 1))
}

export const getPostsWithMarkdown = () =>
  getPosts().map((post) => getPost(post.slug))
