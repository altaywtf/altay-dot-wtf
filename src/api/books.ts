import { readBooksJSON } from 'scripts/books/lib/booksJSON'
import { readMarkdownFile } from 'utils/md'

export type { Book } from 'scripts/books/lib/types'

export const getBooks = () => readBooksJSON().books

export const getBook = (slug: string) => {
  const book = readBooksJSON().books.find((book) => book.slug === slug)

  if (!book) {
    throw new Error(`Book not found: ${slug}.`)
  }

  return {
    book,
    markdown: readMarkdownFile(`${book.notes.url}.md`),
  }
}

export const getBooksWithMarkdown = () => getBooks().map((book) => getBook(book.slug))
