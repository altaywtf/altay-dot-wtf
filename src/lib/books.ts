import { readMarkdownFile } from "@/lib/utils/md";
import { readBooksJSON } from "@/scripts/books/lib/booksJSON";

export type { Book } from "@/scripts/books/lib/types";

export const getBooks = () => readBooksJSON().books;

export const getBook = (slug: string) => {
  const book = readBooksJSON().books.find((book) => book.slug === slug);

  if (!book) {
    throw new Error(`Book not found: ${slug}.`);
  }

  return {
    book,
    markdown: readMarkdownFile(`${book.notes.url}.md`),
  };
};

export const getBooksWithMarkdown = () =>
  getBooks().map((book) => getBook(book.slug));
