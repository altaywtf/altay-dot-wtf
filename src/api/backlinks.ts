import { hasLink } from 'utils/md'
import { getPostsWithMarkdown } from './posts'
import { getBooksWithMarkdown } from './books'

export type Backlink = {
  type: 'post' | 'book'
  title: string
  url: string
}

export const getBacklinks = (url: string): Backlink[] => {
  const posts: Backlink[] = getPostsWithMarkdown()
    .filter(({ markdown }) => hasLink(markdown, url))
    .map(({ post }) => ({
      type: 'post',
      title: post.title,
      url: post.url,
    }))

  const books: Backlink[] = getBooksWithMarkdown()
    .filter(({ markdown }) => hasLink(markdown, url))
    .map(({ book }) => ({
      type: 'book',
      title: `${book.title} by ${book.authors.join(', ')}`,
      url: book.notes.url,
    }))

  return [...posts, ...books]
}
