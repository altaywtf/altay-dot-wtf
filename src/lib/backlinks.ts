import { getBooksWithMarkdown } from '@/lib/books'
import { getPostsWithMarkdown } from '@/lib/posts'
import { hasLink } from '@/lib/utils/md'

export type Backlink = {
  title: string
  type: 'book' | 'post'
  url: string
}

export const getBacklinks = (url: string): Backlink[] => {
  const posts: Backlink[] = getPostsWithMarkdown()
    .filter(({ markdown }) => hasLink(markdown, url))
    .map(({ post }) => ({
      title: post.title,
      type: 'post',
      url: post.url,
    }))

  const books: Backlink[] = getBooksWithMarkdown()
    .filter(({ markdown }) => hasLink(markdown, url))
    .map(({ book }) => ({
      title: `${book.title} by ${book.authors.join(', ')}`,
      type: 'book',
      url: book.notes.url,
    }))

  return [...posts, ...books]
}
