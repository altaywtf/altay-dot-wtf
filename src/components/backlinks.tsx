import { getBacklinks } from "@/lib/backlinks";
import Link from "next/link";
import { BacklinkScroller } from "./backlink-scroller";

const BACKLINK_SOURCE_QUERY_PARAM = "backlinkSource";

export const Backlinks: React.FC<{
  path: string;
}> = ({ path }) => {
  const backlinks = getBacklinks(path);

  return (
    <>
      <BacklinkScroller
        backlinkSourceQueryParam={BACKLINK_SOURCE_QUERY_PARAM}
      />

      {backlinks.length > 0 && (
        <div className="flex flex-col gap-2 rounded bg-neutral-900 p-4">
          <h4 className="text-sm font-semibold">Backlinks</h4>

          <ul className="list-outside list-disc pl-4 space-y-1">
            {backlinks.map((backlink) => (
              <li key={backlink.path} className="text-sm">
                <Link
                  className="hover:underline"
                  href={`${backlink.path}?${BACKLINK_SOURCE_QUERY_PARAM}=${path}`}
                >
                  {backlink.title}
                </Link>
              </li>
            ))}
          </ul>
        </div>
      )}
    </>
  );
};
