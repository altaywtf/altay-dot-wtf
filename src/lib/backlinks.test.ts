import { getBooksWithMarkdown } from "@/lib/books";
import { getPostsWithMarkdown } from "@/lib/posts";
import { getBacklinks } from "./backlinks";

jest.mock("@/lib/posts", () => ({
  getPostsWithMarkdown: jest.fn(),
}));

jest.mock("@/lib/books", () => ({
  getBooksWithMarkdown: jest.fn(() => []),
}));

describe("getBacklinks", () => {
  beforeAll(() => {
    const mockGetPostsWithMarkdown =
      getPostsWithMarkdown as jest.MockedFunction<typeof getPostsWithMarkdown>;

    mockGetPostsWithMarkdown.mockImplementation(() => [
      {
        markdown: "hello and [hi](/posts/hi)",
        post: {
          date: "2020-12-21T20:41:04Z",
          oneliner: "yes",
          readingTime: "1 min read",
          slug: "yes",
          title: "yes",
          url: "/posts/yes",
        },
      },
      {
        markdown: "hello and hi",
        post: {
          date: "2020-12-21T20:41:04Z",
          oneliner: "no",
          readingTime: "1 min read",
          slug: "no",
          title: "no",
          url: "/posts/no",
        },
      },
    ]);

    const mockGetBooksWithMarkdown =
      getBooksWithMarkdown as jest.MockedFunction<typeof getBooksWithMarkdown>;

    mockGetBooksWithMarkdown.mockImplementation(() => [
      {
        book: {
          authors: ["Jess Henderson"],
          coverImage: {
            aspectRatio: 0.8067226890756303,
            blurhash:
              "data:image/jpeg;base64,/9j/2wBDAAYEBQYFBAYGBQYHBwYIChAKCgkJChQODwwQFxQYGBcUFhYaHSUfGhsjHBYWICwgIyYnKSopGR8tMC0oMCUoKSj/2wBDAQcHBwoIChMKChMoGhYaKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCj/wAARCAAEAAMDASIAAhEBAxEB/8QAFQABAQAAAAAAAAAAAAAAAAAAAAX/xAAeEAABAwQDAAAAAAAAAAAAAAADAAECBAUGEQdTY//EABUBAQEAAAAAAAAAAAAAAAAAAAAB/8QAFBEBAAAAAAAAAAAAAAAAAAAAAP/aAAwDAQACEQMRAD8As3jjpg1rDjlmVO0Qi1urH1x80RFB/9k=",
            url: "/images/books/offline-matters/cover.png",
          },
          dateRead: "2021-07-16",
          identifiers: [
            {
              identifier: "9063695780",
              type: "ISBN_10",
            },
            {
              identifier: "9789063695781",
              type: "ISBN_13",
            },
          ],
          notes: {
            url: "/books/offline-matters",
          },
          quote:
            "We are all bored and everything is boring. Occupied 24/7, doing nothing at all.",
          rating: 4,
          remoteCoverImage: {
            url: "https://cdn.webshopapp.com/shops/71491/files/331301478/jess-henderson-offline-matters.jpg",
          },
          slug: "offline-matters",
          title: "Offline Matters",
        },
        markdown: "hello and [hi](/posts/hi)",
      },
    ]);
  });

  it("works", () => {
    const input = "/posts/hi";
    expect(getBacklinks(input)).toMatchSnapshot();
  });
});
