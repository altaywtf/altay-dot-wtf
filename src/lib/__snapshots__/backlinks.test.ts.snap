// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBacklinks works 1`] = `
[
  {
    "title": "yes",
    "type": "post",
    "url": "/posts/yes",
  },
  {
    "title": "Offline Matters by Jess Henderson",
    "type": "book",
    "url": "/books/offline-matters",
  },
]
`;
